### PR TITLE
ProgramService Implemented

### PR DESCRIPTION
**Pull request by kendrickt**
### Description
Implements the ProgramService. 

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Approvals
Can I get one

### CODE REVIEW

**Comment by kendrickt** [comment]
@carolinedanzi I don't know how to generate block ids. Right now it is hard-coded to 1L and are def not unique within a program, let alone across programs. Though I don't think it is necessary/helpful to have them unique across programs.

File: universal-application-tool-0.0.1/app/models/Program.java [hunk 1]
```diff
@@ -39,17 +42,20 @@ public Program(ProgramDefinition definition) {
...
     this.id = this.programDefinition.id();
     this.name = this.programDefinition.name();
     this.description = this.programDefinition.description();
     this.blockDefinitions = mapper.writeValueAsString(this.programDefinition.blockDefinitions());
   }
 
-  /** Populates {@link programDefinition} from column values. */
+  /** Populates {@link ProgramDefinition} from column values. */
   @PostLoad
+  @PostPersist
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/models/Program.java:57]: question: What is the difference between @PostLoad and @PostPersist and why do we need both? Could we document that decision?

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java [hunk 2]
```diff
@@ -23,8 +23,8 @@ public ProgramRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext execu
   }
 
   /** Return all programs in a set. */
-  public CompletionStage<Set<Program>> listPrograms() {
-    return supplyAsync(() -> ebeanServer.find(Program.class).findSet(), executionContext);
+  public CompletionStage<List<Program>> listPrograms() {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/repository/ProgramRepository.java:26]: question: Why the change to list instead of set? I tend to prefer sets if there is no strong reason for maintaining order

nit: If we decide to go with a list, please update comment

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java [hunk 3]
```diff
@@ -33,16 +33,45 @@ public ProgramRepository(EbeanConfig ebeanConfig, DatabaseExecutionContext execu
...
+        },
+        executionContext);
+  }
+
+  public Program updateProgramSync(Program program) {
+    ebeanServer.update(program);
+    return program;
+  }
+
+  public CompletionStage<Program> insertOrUpdateProgram(Program program) {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/repository/ProgramRepository.java:64]: suggestion: I think the only path we need to have async is getting the program. All others are admin-only operations and can be synchronous

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java [hunk 4]
```diff
@@ -2,12 +2,15 @@
...
 import com.google.common.collect.ImmutableList;
+import javax.annotation.Nullable;
+import models.Program;
 
 @AutoValue
 public abstract class ProgramDefinition {
 
   /** Unique identifier for a ProgramDefinition. */
-  public abstract long id();
+  @Nullable
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:12]: suggestion (blocking): Please use Optional if this can be null

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java [hunk 5]
```diff
@@ -38,19 +39,23 @@
...
    * Update a {@link BlockDefinition} with a set of questions.
    *
    * @param programId the ID of the program to update
    * @param blockDefinitionId the ID of the block to update
-   * @param questionDefinitionIds an {@link ImmutableList} of questions for the block
+   * @param questionDefinitions an {@link ImmutableList} of questions for the block
    * @return the updated {@link ProgramDefinition}
+   * @throws ProgramNotFoundException when programId does not correspond to a real Program.
    */
   ProgramDefinition setBlockQuestions(
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramService.java:56]: suggestion (non-blocking): I think eventually we'll want to use the ErrorAnd class here, so we can return validation errors (for example, if there is no block with this ID for this program)

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 import com.google.common.collect.ImmutableList;
 import java.util.Optional;
+import services.question.QuestionDefinition;
 
 /** Operations you can perform on {@link ProgramDefinition}s. */
 public interface ProgramService {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramService.java:8]: suggestion (non-blocking): Eventually we'll want an async version of get program, but that can be in a different PR

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    Program program = new Program(programDefinition);
+    return programRepository.insertProgramSync(program).getProgramDefinition();
+  }
+
+  @Override
+  public ProgramDefinition addBlockToProgram(
+      long programId, String blockName, String blockDescription) throws ProgramNotFoundException {
+    BlockDefinition blockDefinition =
+        BlockDefinition.builder()
+            .setId(1L) // How to autoincrement this block id?
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:56]: Oof that's a good question. Since we aren't relying on a database to do this for us, we may need a class for this that can compute the next available ID

Comment (carolinedanzi) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:56]: Or we can generate unique identifiers based on the program ID and the block name or some other identifying information (maybe initial index?)

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  int getBlockDefinitionIndex(ProgramDefinition programDefinition, Long blockDefinitionId) {
+    return programDefinition.blockDefinitions().stream()
+        .map(b -> b.id())
+        .collect(ImmutableList.toImmutableList())
+        .indexOf(blockDefinitionId);
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:123]: optional suggestion: another way so that you don't have to collect all the IDs:

```java
programDefinition.blockDefinitions().stream()
    .filter(b -> b.id().equals(blockDefinitionId))
    .findFirst();
```

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+  private ProgramDefinition updateProgramDefinitionWithBlockDefinition(
+      ProgramDefinition programDefinition,
+      int blockDefinitionIndex,
+      BlockDefinition blockDefinition) {
+    List<BlockDefinition> mutableBlockDefinitions =
+        new ArrayList<>(programDefinition.blockDefinitions());
+    mutableBlockDefinitions.set(blockDefinitionIndex, blockDefinition);
+    ImmutableList<BlockDefinition> updatedBlockDefinitions =
+        mutableBlockDefinitions.stream().collect(ImmutableList.toImmutableList());
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:134]: nifty trick: `ImmutableList.copyOf(mutableBlockDefinitions);`

File: universal-application-tool-0.0.1/test/repository/WithResettingPostgresContainer.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import io.ebean.Ebean;
+import io.ebean.EbeanServer;
+import models.Applicant;
+import models.Person;
+import models.Program;
+import models.Question;
+import org.junit.Before;
+import play.db.ebean.EbeanConfig;
+
+public class WithResettingPostgresContainer extends WithPostgresContainer {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/test/repository/WithResettingPostgresContainer.java:12]: question: Is there any reason we now need both of these? Or can we reset the other WithPostgresContainer and just have the one?

File: universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.Locale;
+import java.util.Optional;
+import org.junit.Before;
+import org.junit.Test;
+import repository.WithResettingPostgresContainer;
+import services.question.QuestionDefinition;
+
+public class ProgramServiceImplTest extends WithResettingPostgresContainer {
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java:16]: Can we add a test where the block with given ID does not exist on the program?

**Review by carolinedanzi — COMMENTED** [review]
> @carolinedanzi I don't know how to generate block ids. Right now it is hard-coded to 1L and are def not unique within a program, let alone across programs. Though I don't think it is necessary/helpful to have them unique across programs.

Great point - we may be able to make a class to handle this for us, or we can somehow generate IDs based off the block data

File: universal-application-tool-0.0.1/app/models/Program.java (hunk 1 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/models/Program.java:57]: Load - read from DB
Persist - save new row to DB
Update - update existing row to DB

We _could_ have a separate one for `@PostPersist` since that one won't need to `setBlockDefinitions` but I think it's cleaner if we just have one for all `@PostX`.

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java (hunk 2 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/repository/ProgramRepository.java:26]: I changed it because the method name has "list".

File: universal-application-tool-0.0.1/app/repository/ProgramRepository.java (hunk 3 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/repository/ProgramRepository.java:64]: Done!

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java (hunk 5 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramService.java:56]: ahh good call!

File: universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java (hunk 4 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramDefinition.java:12]: Done

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 7 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:56]: Addressed in chat.

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 8 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:123]: I need the index. Is there a better way to get the index?

File: universal-application-tool-0.0.1/test/repository/WithResettingPostgresContainer.java (hunk 10 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/test/repository/WithResettingPostgresContainer.java:12]: the `FunctionalTest` requires state. I think that's not good, but that's the way it is. We can address that later.

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 9 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:134]: Done!

File: universal-application-tool-0.0.1/app/services/program/ProgramService.java (hunk 5 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramService.java:56]: Let's make this a separate PR.

File: universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java (hunk 11 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java:16]: Done

Comment (kendrickt) [reply universal-application-tool-0.0.1/test/services/program/ProgramServiceImplTest.java:16]: Done

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+import repository.ProgramRepository;
+import services.question.QuestionDefinition;
+
+public class ProgramServiceImpl implements ProgramService {
+
+  private ProgramRepository programRepository;
+
+  @Inject
+  public ProgramServiceImpl(ProgramRepository programRepository) {
+    this.programRepository = programRepository;
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:20]: nit: use `checkNotNull`

File: universal-application-tool-0.0.1/app/models/Program.java [hunk 13]
```diff
@@ -29,11 +30,20 @@ public ProgramDefinition getProgramDefinition() {
 
   public Program(ProgramDefinition definition) {
     this.programDefinition = definition;
```
Comment (bion) [inline universal-application-tool-0.0.1/app/models/Program.java:32]: nit: checkNotNull

Comment (kendrickt) [reply universal-application-tool-0.0.1/app/models/Program.java:32]: done

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 12 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:20]: done

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
+
+  int getBlockDefinitionIndex(ProgramDefinition programDefinition, Long blockDefinitionId)
+      throws ProgramBlockNotFoundException {
+    int index =
+        programDefinition.blockDefinitions().stream()
+            .map(b -> b.id())
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:143]: nit: could use method reference `BlockDefinition::id`

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+            .collect(ImmutableList.toImmutableList())
+            .indexOf(blockDefinitionId);
+    if (index == -1) {
+      throw new ProgramBlockNotFoundException(programDefinition.id(), blockDefinitionId);
+    }
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:148]: nit: please add a newline before and after the multiline `if` statement 🙏🏻

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    BlockDefinition blockDefinition =
+        programDefinition
+            .blockDefinitions()
+            .get(blockDefinitionIndex)
+            .toBuilder()
+            .setOptionalPredicate(Optional.of(predicate))
+            .build();
+    return updateProgramDefinitionWithBlockDefinition(
+        programDefinition, blockDefinitionIndex, blockDefinition);
+  }
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:137]: nit: for the methods between L61 and here would you please add some blank lines in the method body to separate some of the statements? It'll make it a lot easier for me to read.

**Review by bion — APPROVED** [review]

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 15 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:152]: done!

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java (hunk 16 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:140]: done!

File: universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+  @Override
+  public ProgramDefinition createProgramDefinition(String name, String description) {
+    Program program = new Program(name, description);
+    return programRepository.insertProgramSync(program).getProgramDefinition();
+  }
+
+  @Override
+  @Transactional
```
Comment (natsid) [inline universal-application-tool-0.0.1/app/services/program/ProgramServiceImpl.java:60]: suggestion: Looking at [these docs](https://ebean.io/docs/transactions/transactional), it seems like to get the benefit of `@Transactional`, we may need to add an `ebean.mf` manifest file with the `transaction-packages` set.

**Review by natsid — COMMENTED** [review]
Good stuff!!
